lmod : support for hierarchical modules

alalazo opened the pull request:
##### Modifications
- [x] extend the list of module file generators to `lmod`
- [x] support for arbitrary virtual packages in module files hierarchies
- [x]  unit tests

##### Related issues / Open questions
  - [TACC/Lmod #114](https://github.com/TACC/Lmod/issues/114)

##### Example
A `modules.yaml` configuration like :
```yaml
modules::
  enable : ['lmod']
  lmod:
    core_compilers: ['gcc@4.8']  # Anything compiled with these compilers will end up in 'Core'
    hierarchical_scheme: ['lapack'] # 'lapack' providers are treated as 'mpi' providers
```
will generate a hierarchical structure that extends the usual `Core/compiler/MPI` structure treating `lapack` providers the same way as `mpi` providers

tgamblin commented on lib/spack/spack/modules.py:
```diff
@@ -259,3 +259,126 @@ def _write(self, m_file):
+            # If the module is MPI parallel then put the modulefile in '<MPI>/<MPI-Version>/<Compiler/<Compiler-Version>'
+            cname = self.spec.compiler.name
+            cversion = self.spec.compiler.version
+            for key, val in self.spec.dependencies.items():
```
@alalazo: You can actually simplify this whole loop to this:
```
if not 'mpi' in spec:
    raise RuntimeError('No MPI dependency found')

mpi_name = spec['mpi'].name
mpi_version = spec['mpi'].version
```
The above is actually more correct, as well, because `spec.dependencies` is just the _direct_ deps.  If you want _all_ dependencies you want to look at `Spec.traverse()`, which will walk the DAG for you.  `Spec.__contains__` does this automatically and tests each dep to see if it satisfies the argument (which is why `'mpi' in spec` works).

Maybe this is better illustrated in an example:
```
>>> from spack.spec import Spec
>>> from spack.graph import graph_ascii

>>> s = Spec('adept-utils')
>>> s.normalize()
>>> graph_ascii(s)
o  adept-utils
|\
o |  mpi
 /
o  boost
>>> 'mpi' in s
True
>>> 'mpich' in s
False
>>> 'openmpi' in s
False

>>> s.concretize()
>>> graph_ascii(s)
o  adept-utils
|\
o |  mpich
 /
o  boost
>>> 'mpi' in s
True
>>> 'mpich' in s
True
>>> 'openmpi' in s
False
```

alalazo replied:
You are right, the loop is definitely not necessary. I also removed the `raise`, as the conditional already checks that `'mpi'` is in the spec.

tgamblin commented on lib/spack/spack/modules.py:
```diff
@@ -259,3 +259,126 @@ def _write(self, m_file):
+                    mpi_version = package.version
+                    break
+            else:
+                raise RuntimeError('No MPI dependency found')
```
@alalazo: I think this will throw an exception if you install anything that doesn't depend on MPI, right?   I guess that needs polishing.

tgamblin commented on lib/spack/spack/modules.py:
```diff
@@ -259,3 +259,126 @@ def _write(self, m_file):
+            hierarchy_name = self._mpi_module_directory(cname, cversion, mpi_name, mpi_version)
+        else:
+            # If I am here it means that I am using the system compiler with some MPI
+            # TODO : decide how to deal with this case
```
We've got external dependencies coming soon...

tgamblin commented on lib/spack/spack/modules.py:
```diff
@@ -259,3 +259,126 @@ def _write(self, m_file):
+
+        :return: True or False
+        """
+        # FIXME: How can I check if a spec has been constructed using the system compiler?
```
Is the system compiler just the OS default compiler?  Spack doesn't really know what that is at the moment, but we could ask the OS or something.  How do you guys typically get this information on your systems?  Is it a default, or is it actually the compiler used to build the OS stack?

alalazo replied:
Yes, it's just the OS default compiler. To organize the modules hierarchically I have to select a default compiler (e.g. `/usr/bin/gcc`) which I will use:
* to build serial libraries that don't need compiler specific optimizations or don't benefit from them 
* to build generic tools, for instance for development purposes (e.g. cmake)
* to bootstrap other compilers if needed

The modules for those things will be placed in `<modules-root>/Core` and will be always visible to the user. On the other hand, modules that are built using a compiler provided in  `<modules-root>/Core` will be placed into their own directory and at most one directory will be visible to users at any time.

tgamblin commented on lib/spack/spack/modules.py:
```diff
@@ -259,3 +259,126 @@ def _write(self, m_file):
+            # If the module is serial and built using a compiler other than the system one,
+            # put the modulefile in '<Compiler>/<Version>'
+            hierarchy_name = self._compiler_module_directory(self.spec.compiler.name, self.spec.compiler.version)
+        elif self._is_mpi_dependent() and not self._use_system_compiler():
```
isn't just `else` sufficient here?

tgamblin commented:
I haven't had a chance to test this but it looks great (and pretty straightforward) to me.  I could see extending this with a syntax to describe different hierarchical schemes (e.g. not just mpi/compiler/etc.).  If you can address some of the comments and push changes, I think this is pretty close to being merged into develop with preliminary support.

I'll probably add something to disable/enable certain types of modules in the config file, as well, since Spack now does three types and I'm not sure everyone wants them all.

One more question: Are there edits that need to be made to `share/spack/setup-env.[c]sh`?  Or are the env vars used by tcl modules and `Lmod` the same?

alalazo commented:
> I could see extending this with a syntax to describe different hierarchical schemes (e.g. not just mpi/compiler/etc.).


If you have already a clear idea in mind, I am available to extend this implementation.


> Are there edits that need to be made to share/spack/setup-env.[c]sh? Or are the env vars used by tcl modules and `Lmod` the same

To my knowledge `Lmod` is an implementation of Environment Modules with some unique features, thus the variables that need to be set are the same. The line:
```
_spack_pathadd MODULEPATH "$_sp_share_dir/modules/$SYS_TYPE"
```
will set-up a configuration where `Lmod` is used with a "flat" module directory. To use hierarchical modulefiles the same line should read:
```
_spack_pathadd MODULEPATH "$_sp_share_dir/lmod/modulefiles/Core"
```
This brings me a couple question:
* Should I change the name of  `class LmodModule` to something like `class LuaHierarchicalModule`? (It may be more appropriate since `Lmod` can also deal with `TclModule`)
* I miss a distinction based on `SYS_TYPE` in my directory name for hierarchical modules. Should it be added?

alalazo commented:
I still need to implement:
```python
@property
def use_name(self):
    # FIXME : this needs to be implemented. It seems to be used only in module.module_find(m_type, spec_array)
    pass
```
What is this function used for?

tgamblin commented:
>> I could see extending this with a syntax to describe different hierarchical schemes (e.g. not just mpi/compiler/etc.).

> If you have already a clear idea in mind, I am available to extend this implementation.

I have some ideas but not enough to constitute a 'spec'.  So let's go with your version for now and I can add some configurability later.

>>Are there edits that need to be made to share/spack/setup-env.[c]sh? Or are the env vars used by tcl modules and Lmod the same

>To my knowledge Lmod is an implementation of Environment Modules with some unique features, thus the variables that need to be set are the same. The line:
>```
>_spack_pathadd MODULEPATH "$_sp_share_dir/modules/$SYS_TYPE"
>```
>will set-up a configuration where Lmod is used with a "flat" module directory. To use hierarchical modulefiles the same line should read:
>```
>_spack_pathadd MODULEPATH "$_sp_share_dir/lmod/modulefiles/Core"
>```

Ok cool -- I think this is fine.  I can add some configurable options to select between the TCL module and Lmod.

>This brings me a couple question:
>
>Should I change the name of class `LmodModule` to something like class `LuaHierarchicalModule`? (It may be more appropriate since Lmod can also deal with TclModule)

I think the first one's fine.  Most people will want Lmod to be hierarchical.

>I miss a distinction based on SYS_TYPE in my directory name for hierarchical modules. Should it be added?

Yes, please.

tgamblin commented:
>I still need to implement:
>```python
>@property
>def use_name(self):
>    # FIXME : this needs to be implemented. It seems to be used only in module.module_find(m_type, spec_array)
>    pass
>```
>What is this function used for?

`use_name` is basically what's used for the filename of the module.  It's also what's printed when you do a `spack module find`, and what gets loaded when you do `spack load`.   Given that you already overrode `file_name`, I think the right thing to do is to move the code in `file_name` to `use_name` so that `use_name` prints the name without a suffix and `file_name` has the suffix...  Basically use_name should print whatever you'd pass to `module load`.

alalazo commented:
I implemented `use_name` and added `self.spec.architecture` to the naming scheme. I am not sure anyhow that there won't be some glitches when trying to load or search hierarchical modules directly from spack.

In fact `Lmod` provides its own command to search for a package:
```
$ module spider hdf5

----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  hdf5:
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
     Versions:
        hdf5/1.8.13

----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  To find detailed information about hdf5 please enter the full name.
  For example:

     $ module spider hdf5/1.8.13
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

$ module spider hdf5/1.8.13

----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  hdf5: hdf5/1.8.13
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

    This module can only be loaded through the following modules:

      gcc/5.2.0  mpich/3.0.4
      mpich/3.0.4
 
    Help:
      HDF5 is a data model, library, and file format for storing and managing data. It supports an unlimited variety of datatypes, and is designed for flexible and efficient I/O and for high volume and complex da
ta. 
```
and sometimes you may access to a module in multiple ways (as in the example above, where hdf5/1.8.13 is compiled using either default compiler over mpich or gcc 5.2.0 over mpich)

boegel commented:
I took a look at this, there's a lot of different things going on here at the same time:

* support for using Lmod as a modules tool
* support for using a hierarchical module naming scheme
* support for using module files in Lua syntax

I'm not sure whether you want to be doing everything at the same time (although the combination sure makes sense).

As @tgamblin already mentioned, this also hard-codes to the traditional `Core`/`Compiler`/`MPI` hierarchy, while having something more flexible would be nice.
In EasyBuild, we implement support the hierarchical modules via a custom module naming scheme, for example: https://github.com/hpcugent/easybuild-framework/blob/master/easybuild/tools/module_naming_scheme/hierarchical_mns.py .
But, you can easily tweake that slightly to your needs, for example by adding an extra hierarchy level for BLAS or whatever, by copying and modifying that module slightly, and then configuring EB to pick it up and use it.

The support for using Lmod as a modules tool (i.e., to use `lmod python <command>` rather than `modulecmd python <command>`), and to generate module files Lua syntax, is handled entirely separately from each other in EasyBuild.
Although you (kind of) need to use Lmod when doing a hierarchy, you don't need to use module files in Lua syntax; you can do it in Tcl syntax just fine if you want to (for example, but avoid 'vendor locking' to using Lmod).
Vice versa, you can also use Lmod without using a module hierarchy.

Since I was involved with much of the support for using Lmod, generating module files in Lua syntax and supporting hierarchical modules in EasyBuild, I'm aware of the hoops you need to make Spack jump through. I'm happy to answer any questions. The paper on using Lmod and a module hierarchy I'm a co-author of is a good read to get a good view on this, see http://hpcugent.github.io/easybuild/files/hust14_paper.pdf .

alalazo commented:
@boegel : many thanks for taking the time to review and comment on this PR.

>I took a look at this, there's a lot of different things going on here at the same time:
> - support for using Lmod as a modules tool
> - support for using a hierarchical module naming scheme
> - support for using module files in Lua syntax
>
> I'm not sure whether you want to be doing everything at the same time (although the combination sure makes sense).

This changeset addresses only two of the three aspects you cited, that is the support for using a hierarchical module naming scheme and for writing module files in Lua syntax.

I merged together those two aspects mainly because:
 1. Lua seems to be the native way to write module files in Lmod
 1. to make full use of the features provided by Lmod you need to extend anyhow the TCL modulefiles syntax with the `family` function (or at least this is how I got it from Lmod documentation)

Do you see relevant use cases where it would it make sense to use Lmod with "traditional" TCL modulefiles, or where one would use `modulecmd` with a hierarchical layout? 

>As @tgamblin already mentioned, this also hard-codes to the traditional Core/Compiler/MPI hierarchy, while having something more flexible would be nice.
In EasyBuild, we implement support the hierarchical modules via a custom module naming scheme, for example: https://github.com/hpcugent/easybuild-framework/blob/master/easybuild/tools/module_naming_scheme/hierarchical_mns.py .
But, you can easily tweake that slightly to your needs, for example by adding an extra hierarchy level for BLAS or whatever, by copying and modifying that module slightly, and then configuring EB to pick it up and use it.

I agree that having more flexibility would be nice, but I don't see an easy way at the moment to go far ahead of what is in this PR. As I see it 'levels' in a hierarchical naming scheme maps directly to virtual dependencies (you have 'compiler', 'mpi', 'blas', etc.). Having the traditional Core/Compiler/MPI hierarchy covers a good part of our use cases, and is pretty straightforward. Adding other things to the mix (like 'blas', 'lapack', 'cuda', etc.) naturally calls for a better support of a dependency matrix, where things should be visible only if providers of all their virtual dependencies are loaded. As far as I understood this is still an open area of research for Lmod. How do you deal in EasyBuild with this case?

> The support for using Lmod as a modules tool (i.e., to use lmod python <command> rather than modulecmd python <command>), and to generate module files Lua syntax, is handled entirely separately from each other in EasyBuild.

I guess module files generation and support for the delegation to a module tool (e.g. `spack load` and similar commands) are separated by design also in `spack`, but @tgamblin may obviously say more about that. I don't provide support for Lmod as a tool in this PR, but I am confident it is something that will be addressed in the near future.

alalazo commented:
@tgamblin Do you think this PR miss anything to be merged?

alalazo commented:
@tgamblin Odd : this old PR seem not to be able to trigger Travis anymore...

adamjstewart commented:
Even when you rebase it?

alalazo commented:
@adamjstewart Yes

alalazo commented:
@tgamblin Any chance this could be reviewed manually ?

adamjstewart commented:
Huh, Travis works again for you too eh?

alalazo commented:
Not really... it seems it's pulling a travis.yaml file from a while ago...
